### PR TITLE
Group saved for backward tensors produced by TransfomerEngine together with all other saved for backward tensors

### PR DESCRIPTION
**Pull request by IvanYashchuk**
TLDR: `Tensor.backward(retain_graph=True)` is a public PyTorch feature that may be used in a lot of code, we should support it. Thunder without TransformerEngine works, but with TransformerEngine there was a bug in Thunder code that is fixed in this PR. There's still a bug on the TransformerEngine's side (https://github.com/NVIDIA/TransformerEngine/issues/990).

This PR also bumps the required minimal version of TransformerEngine to 1.8+.

Previously we were hiding "saved_tensors" inside an opaque context object that TransformerEngine populates inside `_Linear.forward` and we were relying on passing a dict representation of this object to be passed from `ThunderFunction.forward` to `ThunderFunction.backward` instead of dealing with it the same way as we deal with all other tensors to be saved for backward here
https://github.com/Lightning-AI/lightning-thunder/blob/e75ddc1353178d5d74fcc5d29154efd8f77605a5/thunder/executors/torch_autograd.py#L73
A dict representation was needed to let Thunder clear it during backward and release the memory occupied by saved by TE tensors. However, we already have a mechanism to clear saved tensors and also that works correctly with PyTorch's `retain_graph=True`, so we should be using it.
For Thunder's general "saved tensor" list to include also TransformerEngine's saved tensors we need to modify the meta functions to model it explicitly like we do in all other forward meta functions.

The code snippet from https://github.com/Lightning-AI/lightning-thunder/issues/701 still doesn't work, but the problem is not in the Thunder code.

Here's how the forward and backward traces changed:
```py
import torch
import thunder
from thunder.executors.transformer_engineex import transformer_engine_ex

def foo(x, w):
    return torch.nn.functional.linear(x, w)
x = torch.randn(16, 16, requires_grad=True, device='cuda')
w = torch.randn(16, 16, requires_grad=True, device='cuda')

fn = thunder.jit(foo, executors=[transformer_engine_ex,])
fn(x, w)

print(thunder.last_traces(fn)[-1])
print(thunder.last_backward_traces(fn)[-1])
```
```diff
def augmented_forward_fn(x, w):
  # x: "cuda:0 f32[16, 16]"
  # w: "cuda:0 f32[16, 16]"
- (t8, ctx_te_1) = te_linear_0(x, w, None)
+  (t8, (t3, t4, t5, _, t6, t7, _), ctx_te_1) = te_linear_0(x, w, None)
- return {'output': t8, 'flat_args': [x, w], 'flat_output': (t8,)}, ((), (ctx_te_1,))
+  return {'output': t8, 'flat_args': [x, w], 'flat_output': (t8,)}, ((t3, t4, t5, t6, t7), (ctx_te_1,))
```
```diff
def backward_fn(saved_for_backward, cotangents):
  # saved_for_backward: "Collection"
  # cotangents: "Collection"
-  _, C1, = saved_for_backward
+  C0, C1, = saved_for_backward
  clear_mutable_collection(saved_for_backward)
  del saved_for_backward
  t9, = cotangents
  clear_mutable_collection(cotangents)
  del cotangents
-
+  t3, t4, t5, t6, t7, = C0
+  clear_mutable_collection(C0)
+  del C0
  ctx_te_3, = C1
  clear_mutable_collection(C1)
  del C1
- (t12, t13, _) = te_functional_linear_backward(t9, (16, 16), (16, 16), None, ctx_te_3)
+  (t12, t13, _) = te_functional_linear_backward((16, 16), (16, 16), None, ctx_te_3, (t3, t4, t5, None, t6, t7, None), t9)
- clear_mutable_collection(ctx_te_1)
+  del ctx_te_3, t3, t4, t5, t6, t7, t9
  te_sync_fp8_meta_bwd()
  return (t12, t13)
```

### CODE REVIEW

**Comment by kshitij12345** [comment]
On TE v1.8 ('1.8.0+37280ec') - I am seeing failure when running - `pytest thunder/tests/test_transformer_engine_executor.py`. Looks like the number of saved tensors has changed. Since, `_Linear` is a non-public API, it's inputs and outputs change frequently. The [PR](https://github.com/NVIDIA/TransformerEngine/pull/707) adding stable functional interface should be merged soon (it is currently in review). Till then we can either add switches based on version and return output accordingly or maybe wait till the mentioned PR is merged and then use the stable functional API.

```python
    def augmented_forward_fn(x, w):
      # x: "cuda:0 f32[16, 16]"
      # w: "cuda:0 f32[16, 16]"
>     (t8, (t3, t4, t5, _, t6, t7, _), ctx_te_19) = te_linear_18(x, w, None)
E     ValueError: not enough values to unpack (expected 7, got 6)
```

**Comment by IvanYashchuk** [comment]
> On TE v1.8 ('1.8.0+37280ec') - I am seeing failure when running - `pytest thunder/tests/test_transformer_engine_executor.py`. Looks like the number of saved tensors has changed. Since, `_Linear` is a non-public API, it's inputs, outputs (and other details) change frequently. The [PR](https://github.com/NVIDIA/TransformerEngine/pull/707) adding stable functional interface should be merged soon (it is currently in review). Till then we can either add switches based on version and return output accordingly or maybe wait till the mentioned PR is merged and then use the stable functional API.
> 
> Failure -
> 
> ```python
>     def augmented_forward_fn(x, w):
>       # x: "cuda:0 f32[16, 16]"
>       # w: "cuda:0 f32[16, 16]"
> >     (t8, (t3, t4, t5, _, t6, t7, _), ctx_te_19) = te_linear_18(x, w, None)
> E     ValueError: not enough values to unpack (expected 7, got 6)
> ```

Thank you, we already have a number of these switches, and I'll add another one.

**Comment by IvanYashchuk** [comment]
Instead of adding a switch I have increased the minimal required version of TransformerEngine to be 1.8 and now the saved tensor in the meta functions models the saving of 6 objects as in TE 1.8 and the current TE main.

**Review by kshitij12345 — APPROVED** [review]
LGTM, have verified that the tests pass on v1.8 (stable), v1.9 and v1.10 (latest main). Thanks @IvanYashchuk 

cc: @t-vi

**Comment by IvanYashchuk** [comment]
Awesome! Thank you for checking!

**Review by t-vi — APPROVED** [review]
Thank you @IvanYashchuk  @kshitij12345
